numpy/_core/src/umath/dispatching.cpp

```diff
@@ -976,18 +1029,15 @@ promote_and_get_ufuncimpl(PyUFuncObject *ufunc,
         }
     }
 
-    PyObject *info;
-    Py_BEGIN_CRITICAL_SECTION((PyObject *)ufunc);
-    info = promote_and_get_info_and_ufuncimpl(ufunc,
+    PyObject *info = try_promote_and_get_info_and_ufuncimpl(ufunc,
```

Comment:
I could also make it so this function only gets called on the GIL-disabled build, which might be slightly faster on the GIL-enabled build

Reply:
The one part that I do not like about this construct is that it is not super obvious that the two pieces of code are identical. Maybe just a longer name, more like `mutex_wrapped_promote_and_get_info_and_ufuncimpl` would be an idea?